deepmd/pt/utils/env_mat_stat.py

```diff
@@ -0,0 +1,206 @@
+            device=env.DEVICE,
+        )
+        for system in data:
+            coord, atype, box, natoms = (
+                system["coord"],
+                system["atype"],
+                system["box"],
+                system["natoms"],
+            )
+            (
+                extended_coord,
+                extended_atype,
+                mapping,
+                nlist,
+            ) = extend_input_and_build_neighbor_list(
+                coord,
+                atype,
+                self.descriptor.get_rcut(),
+                self.descriptor.get_sel(),
+                distinguish_types=self.descriptor.distinguish_types(),
+                box=box,
+            )
+            env_mat, _, _ = prod_env_mat_se_a(
+                extended_coord,
+                nlist,
+                atype,
+                zero_mean,
+                one_stddev,
+                self.descriptor.get_rcut(),
+                # TODO: export rcut_smth from DescriptorBlock
+                self.descriptor.rcut_smth,
+            )
+            env_mat = env_mat.view(
+                coord.shape[0], coord.shape[1], self.descriptor.get_nsel(), 4
+            )
+
+            if "real_natoms_vec" not in system:
+                end_indexes = torch.cumsum(natoms[0, 2:], 0)
+                start_indexes = torch.cat(
+                    [
+                        torch.zeros(1, dtype=torch.int32, device=env.DEVICE),
+                        end_indexes[:-1],
+                    ]
+                )
+                for type_i in range(self.descriptor.get_ntypes()):
+                    dd = env_mat[
+                        :, start_indexes[type_i] : end_indexes[type_i], :, :
+                    ]  # all descriptors for this element
+                    env_mats = {}
+                    env_mats[f"r_{type_i}"] = dd[:, :, :, :1]
+                    env_mats[f"a_{type_i}"] = dd[:, :, :, 1:]
+                    yield self.compute_stat(env_mats)
+            else:
+                for frame_item in range(env_mat.shape[0]):
+                    dd_ff = env_mat[frame_item]
+                    atype_frame = atype[frame_item]
+                    for type_i in range(self.descriptor.get_ntypes()):
+                        type_idx = atype_frame == type_i
+                        dd = dd_ff[type_idx]
+                        dd = dd.reshape([-1, 4])  # typen_atoms * nnei, 4
+                        env_mats = {}
+                        env_mats[f"r_{type_i}"] = dd[:, :1]
+                        env_mats[f"a_{type_i}"] = dd[:, 1:]
+                        yield self.compute_stat(env_mats)
+
+    def get_hash(self) -> str:
+        """Get the hash of the environment matrix.
+
+        Returns
+        -------
+        str
+            The hash of the environment matrix.
+        """
+        return get_hash(
+            {
+                "type": "se_a",
+                "ntypes": self.descriptor.get_ntypes(),
```

Comment:
Descriptors that have different `type_map` but the same ntypes will get the same hash, which will mismatch the stat.

Reply:
I see. Since a descriptor doesn't have `type_map`, I will not add it here but add it in the model level.

Reply:
Fixed in 6a83465c7447a5aa159bbb892cc4b2b08e4dc843